cdap-app-fabric/src/main/java/co/cask/cdap/app/store/RuntimeStore.java

```diff
@@ -0,0 +1,123 @@
+import co.cask.cdap.proto.WorkflowNodeStateDetail;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Manages the runtime lifecycle of a {@link Program}.
+ */
+public interface RuntimeStore {
+
+  /**
+   * Compare and set operation that allow to compare and set expected and update status.
+   * Implementation of this method should guarantee that the operation is atomic or in transaction.
+   *
+   * @param id              Info about program
+   * @param pid             The run id
+   * @param expectedStatus  The expected value
+   * @param updateStatus    The new value
+   */
+  void compareAndSetStatus(Id.Program id, String pid, ProgramRunStatus expectedStatus, ProgramRunStatus updateStatus);
+
+  /**
+   * Logs start of program run.
+   *
+   * @param id         Info about program
+   * @param pid        run id
+   * @param startTime  start timestamp in seconds; if run id is time-based pass the time from the run id
+   * @param twillRunId twill run id
+   * @param runtimeArgs the runtime arguments for this program run
+   * @param systemArgs the system arguments for this program run
+   */
+  void setStart(Id.Program id, String pid, long startTime, @Nullable String twillRunId,
+                Map<String, String> runtimeArgs, Map<String, String> systemArgs);
+
+  /**
+   * Logs start of program run. This is a convenience method for testing, actual run starts should be recorded using
+   * {@link #setStart(Id.Program, String, long, String, Map, Map)}.
+   *
+   * @param id        Info about program
+   * @param pid       run id
+   * @param startTime start timestamp in seconds; if run id is time-based pass the time from the run id
+   */
+  @VisibleForTesting
+  void setStart(Id.Program id, String pid, long startTime);
+
+  /**
+   * Logs end of program run.
+   *
+   * @param id      id of program
+   * @param pid     run id
+   * @param endTime end timestamp in seconds
+   * @param runStatus   {@link ProgramRunStatus} of program run
+   */
+  void setStop(Id.Program id, String pid, long endTime, ProgramRunStatus runStatus);
+
+  /**
+   * Logs end of program run.
+   *
+   * @param id      id of program
+   * @param pid     run id
+   * @param endTime end timestamp in seconds
+   * @param runStatus   {@link ProgramRunStatus} of program run
+   * @param failureCause failure cause if the program failed to execute
+   */
+  void setStop(Id.Program id, String pid, long endTime, ProgramRunStatus runStatus,
+               @Nullable BasicThrowable failureCause);
+
+  /**
+   * Logs suspend of a program run.
+   * @param id      id of the program
+   * @param pid     run id
+   */
+  void setSuspend(Id.Program id, String pid);
+
+  /**
+   * Logs resume of a program run.
+   * @param id      id of the program
+   * @param pid     run id
+   */
+  void setResume(Id.Program id, String pid);
```

Comment:
I've already have changed it in a separate branch, to use the new classes, but for simplicity, I will open that as a separate PR (pending item #3 in this PR description).